<img src="http://sonetodigital.com.br/adm/wp-content/uploads/2018/12/logo-fiap.png" width="150" align="left"/>

<br/><br/>

<font size=5 color="#ed145b"><b> Engenharia da Computação </b></font><font size=5 color="#ed145b"> - Inteligência Artificial e computacional </font>

<font size=4 color="#ed145b"> Profa. Thais Rodrigues Neubauer - profthais.neubauer@fiap.com.br</font>

<font size=5 ><b>Aula 01</b></font>

## Introdução a Python

Abra o nosso repositório no [Binder](https://mybinder.org/v2/gh/thaisneubauer/FIAP-3ECA-IA/master)

Duas das principais ferramentas para realizar manipular, tratar, sumarizar e visualizar dados em Python são **Numpy** e **Pandas**. São bibliotecas (conjuntos de classes, funções e objetos) feitos para análise de dados. 

Enquanto Numpy foca em manipulação de vetores e matrizes, muito úteis para realizar operações de álgebra linear (como decomposição de matrizes, calculo de similaridade e distâncias, etc), o Pandas é focado em dados tabulares e manipulações de dados estruturados.

In [1]:
import pandas as pd
import numpy as np

### Numpy e listas em Python

In [2]:
lista_de_ints = [1, 2, 3, 4]
lista_de_strings = ["um", "dois"]
lista_de_listas = [[1,2], [1,2,3,4]]
outra_lista = [5, 5, "banana", 8]

Listas são pouco indicadas para fazer contas. 

Suponha que tenhamos duas listas para gerênciar uma compra em uma loja: uma contendo a quantidade de itens, outra o valor dos ítens.

In [3]:
quantidades = [1, 10, 0, 5, 2]
precos = [10.5, 20.3, 100.75, 0.5, 2.0]

<div class="alert alert-block alert-info">
<b>Exercício:</b>
Calcule o preço total da compra do cliente, multiplicando a quantidade pedida do primeiro produto pelo seu preço, a quantidade pedida do segundo produto pelo seu preço e assim por diante.
</div>

<div class="alert alert-block alert-success">
<b>Solução:</b>
</div>

In [4]:
resultado = 0
for i in range(len(precos)):
    resultado += quantidades[i] * precos[i]
        
resultado

220.0

Alternativamente, usando a função `sum` que soma todos os itens de uma lista:

In [5]:
sum([q * p for p, q in zip(quantidades, precos)])

220.0

<div class="alert alert-block alert-danger">
<b>Não funciona:</b>
</div>

In [6]:
#pagtos = quantidades * precos

Soma de listas é o mesmo que *concatenação* (listas padrão do Python)

In [7]:
quantidades + precos

[1, 10, 0, 5, 2, 10.5, 20.3, 100.75, 0.5, 2.0]

Com Numpy...

In [8]:
x = np.array([1, 2, 3])
y = np.array([5.2, 3.4, 1.0])
x + y

array([6.2, 5.4, 4. ])

In [9]:
x * y

array([5.2, 6.8, 3. ])

<div class="alert alert-block alert-info">
<b>Exercício:</b>

Calcule o preço total da compra do cliente, multiplicando a quantidade pedida do primeiro produto pelo seu preço, a quantidade pedida do segundo produto pelo seu preço e assim por diante. Dessa vez, faça uso dos arrays do numpy.
</div>

In [10]:
quantidades_np = np.array([1, 10, 0, 5, 2])
precos_np = np.array([10.5, 20.3, 100.75, 0.5, 2.0])

In [11]:
quantidades_np

array([ 1, 10,  0,  5,  2])

<div class="alert alert-block alert-success">
<b>Solução:</b>
</div>

In [12]:
pagtos = quantidades_np * precos_np
pagtos.sum()

220.0

Alternativamente

In [13]:
(quantidades_np * precos_np).sum()

220.0

#### Acessando e mudando elementos do array

Algumas vezes, precisamos acessar (por qualquer motivo que seja) um sub-conjunto de um array do numpy. Pode ser que queiramos usar apenas um elemento (ex.: precisamos mandar para alguém apenas o preço do segundo item). Outras vezes, precisamos de um pedaço maior do array, em outras palavras, um sub-array.

Para acessar apenas um elemento do array, podemos usar a sintaxe `a[i]`. No caso dos preços, por exemplo, podemos acessar o preço do terceiro item fazendo:

In [14]:
precos[2]

100.75

<div class="alert alert-block alert-warning">
<b>Atenção:</b>

Assim como nas listas, a contagem dos índices começa de zero. Ou seja, `0` é a posição do primeiro elemento, `1` é a posição do segundo etc.
</div>

Podemos também usar números negativos para indicar os elementos do final para o início da lista (direita para esquerda); dessa forma, acessamos o último elemento com `-1`, o penúltimo com `-2` e assim por diante.

In [15]:
precos[-1] # equivalente a precos[len(precos)-1]

2.0

Podemos ainda usar listas para acessar multiplos elementos de uma vez, fazendo `a[[1,2,3,4]]`. Por exemplo, se quisermos os preços dos três primeiros itens, podemos fazer: 

In [16]:
precos_np[[0,1,2]]

array([ 10.5 ,  20.3 , 100.75])

Para acessar elementos contíguos (ou seja, adjacentes) como no caso acima, podemos também usar o que é chamado de `slicing`. Slices são escritos como `i:j` e significam algo como "acessar os itens da posição `i` (inclusa) até a posição `j` (não inclusa)". Por exemplo, para pegar os preços dos três primeiros itens, faríamos:

In [17]:
precos[0:3]

[10.5, 20.3, 100.75]

In [18]:
precos_np[0:3]

array([ 10.5 ,  20.3 , 100.75])

<div class="alert alert-block alert-info">
<b>Exercício:</b>

Sobre o array `x`, encontre:
    
* O décimo terceiro elemento
* Um sub-array com o primeiro, terceiro e décimo elemento
* Um sub-array com os 5 primeiros elementos
* Um sub-array com os 3 últimos elementos
</div>

In [19]:
x = np.random.randint(0, 100, size=30)

In [20]:
x

array([96, 26, 90, 50, 47, 67, 69, 27, 14, 65, 10, 45, 83, 84, 76, 61, 18,
       34, 86, 39, 55, 39, 78, 74, 39, 13, 15, 42, 78, 56])

<div class="alert alert-block alert-success">
<b>Solução:</b>
</div>

In [21]:
x[12]

83

In [22]:
x[[0, 2, 9]]

array([96, 90, 65])

In [23]:
x[0:5]

array([96, 26, 90, 50, 47])

In [24]:
x[-3:]

array([42, 78, 56])

#### Máscaras

Para acessar elementos de um array, também podemos usar uma lista de Booleanos (`True` ou `False`) do mesmo tamanho do array. Esta lista é chamada de máscara.

A máscara representa quais elementos você quer preservar (representado por `True`) e quais você quer descartar (`False`) do seu resultado final. Por exemplo, para acessar os preços do primeiro e do terceiro item, fazemos:

In [25]:
precos_np[[True, False, True, False, False]]

array([ 10.5 , 100.75])

<div class="alert alert-block alert-warning">
<b>Atenção:</b>

Você também pode usar um np.array de booleanos se desejar. Usamos a lista por simplicidade.
</div>

O método da máscara pode parecer inútil, mas é muito útil quando você quer filtrar elementos de um array. Por exemplo, imagine que queremos somente os preços maiores que 10:

In [26]:
mascara = precos_np > 10

In [27]:
# A máscara é um np.array de booleanos
mascara

array([ True,  True,  True, False, False])

In [28]:
# Usamos então a máscara para acessar os elementos que queremos
precos_np[mascara]

array([ 10.5 ,  20.3 , 100.75])

<div class="alert alert-block alert-info">
<b>Exercício:</b>

Alguns produtos tiveram mais de uma unidade pedida pelo usuário. Encontre os preços desse produtos.
</div>

<div class="alert alert-block alert-success">
<b>Solução:</b>
</div>

In [29]:
mascara = quantidades_np > 1
precos_np[mascara]

array([20.3,  0.5,  2. ])

#### Alterando elementos do array

Se você precisar, pode alterar elementos de um array. Para fazer isso é muito simples, para mudar o preço do terceiro para 90, fazemos:

In [30]:
precos[2] = 90

In [31]:
precos

[10.5, 20.3, 90, 0.5, 2.0]

Mais do que isso, você pode alterar multiplos elementos de uma vez! Imagine que você queira dar um desconto de 50% nos três primeiros itens da lista. Basta fazer:

In [32]:
precos_np[0:3] = precos_np[0:3] / 2

In [33]:
precos_np

array([ 5.25 , 10.15 , 50.375,  0.5  ,  2.   ])

Se você não tiver uma cópia dos dados originais guardada em algum outro local, você pode preferir criar uma cópia dessa lista de precos para depois alterar os valores na cópia e, assim, não perder os dados originais. Para isso, é necessário utilizar a função `copy`:

In [34]:
precos_np = np.array([10.5, 20.3, 100.75, 0.5, 2.0])

In [35]:
precos_np_promocao = precos_np.copy()
precos_np_promocao[0:3] = precos_np_promocao[0:3] / 2

In [36]:
precos_np

array([ 10.5 ,  20.3 , 100.75,   0.5 ,   2.  ])

In [37]:
precos_np_promocao

array([ 5.25 , 10.15 , 50.375,  0.5  ,  2.   ])

#### Numpy Arrays de multiplas dimensões

![title](../images/arrays.png)

Numpy dá ferramentas excelentes para a manipulação de vetores (no sentido matemático). No entanto, também é ideal quando lidamos com dados de dimensionalidade mais alta, como matrizes ou arrays em 3D. Matrizes, por serem bidimensionais (altura e largura), são capazer de representar objetos como rotações e imagens preto e branco. Arrays 3D podem ser usados para representar imagens coloridas (onde a terceira dimensão é a cor), ou um vídeo (onde a terceira dimensão é o tempo).

Para criar uma matriz, basta fazer:

In [38]:
m1 = np.array([[1,2,3], [4,5,6]])

In [39]:
m1

array([[1, 2, 3],
       [4, 5, 6]])

Podemos também criar uma matriz aleatória, fazendo:

In [40]:
m2 = np.random.randint(0, 10, size=(3, 2))

In [41]:
m2

array([[8, 6],
       [3, 5],
       [0, 9]])

Como são matrizes, podemos fazer operações entre elas, como por exemplo a multiplicação matricial:

In [42]:
np.dot(m1, m2)

array([[ 14,  43],
       [ 47, 103]])

Você também consegue transpor uma matriz, fazendo simplesmente:

In [43]:
m1.T

array([[1, 4],
       [2, 5],
       [3, 6]])

Incrível! Diversas operação estão disponíveis no `numpy` e também em outras bibliotecas como `scipy`.

#### Acessando dados

Para acessar os dados, basta usar `matriz[linha, coluna]` onde `linha` é a posição da linha e `coluna` é a posição da coluna. Por exemplo, para acessarmos a primeira linha e segunda coluna de uma matriz, fazemos:

In [44]:
m1[0, 1]

2

Técnicas para acessar dados que funcionam para arrays de 1 dimensão, também funcionam aqui. Um exemplo é usar slices:

In [45]:
m1[0:2, 0:2]

array([[1, 2],
       [4, 5]])

E também máscaras. Por exemplo, podemos querer só as linhas da matriz cuja segunda coluna tem valor superior a zero:

In [46]:
mascara = m2[:, 1] > 0  # mascara das linhas cuja segunda coluna é superior a zero
m2[mascara, :]

array([[8, 6],
       [3, 5],
       [0, 9]])

<div class="alert alert-block alert-info">
<b>Exercício:</b>

Crie uma matriz do numpy com números aleatórios de 0 a 100. O tamanho da matriz deve ser 100 linhas por 50 colunas. Depois, use a técnica da máscara para selecionar só as linhas da matriz cujo valor da terceira coluna seja maior do que o da primeira coluna. Quantas linhas sobraram?
</div>

<div class="alert alert-block alert-success">
<b>Solução:</b>
</div>

In [47]:
matriz = np.random.randint(0, 100, size=(100, 50))

In [48]:
mascara = matriz[:, 2] > matriz[:, 0]
matriz[mascara, :]

array([[ 2, 73, 36, ..., 71, 96, 25],
       [49,  8, 54, ..., 18, 65, 37],
       [14, 47, 19, ..., 40, 94, 92],
       ...,
       [34, 30, 71, ..., 10, 65, 54],
       [66, 30, 87, ...,  8, 46, 93],
       [13, 93, 73, ..., 60, 12, 65]])

#### Operações de agregação

Com numpy, você consegue agregar dados com extrema facilidade. Por exemplo, é possível calcular soma, média, desvio padrão, etc. de uma matrix por linha, or coluna ou entre todos os elementos. Essas operações facilitam muito a computação de estatisticas de seus dados.

Por exemplo, muitos dos modelos de machine learning funcionam melhor se os dados estiverem normalizadas, ou seja, subtraído da sua média e divido pela variância. Isso (usualmente, dependendo da distribuição original) faz com que os dados tenham a média 0 e o desvio padrão 1.

![standardization](../images/standardization.jpg)

 Em numpy, imagine que temos uma matriz cujas linhas representam pessoas, e uma coluna representa a altura (cms) e a outra coluna o peso da pessoa (em kgs). Para normalizar as alturas e os pesos, você poderia fazer:

In [49]:
# Criando alturas e pesos aleatórios 
alturas_pesos = np.concatenate([np.random.randint(140, 200, size=(10,1)), np.random.randint(60, 110, size=(10, 1))], axis=1)
alturas_pesos

array([[187,  73],
       [172,  80],
       [146,  92],
       [172,  98],
       [156,  68],
       [145,  61],
       [158,  85],
       [177,  97],
       [167,  82],
       [189,  60]])

In [50]:
# normalizando
(alturas_pesos - alturas_pesos.mean(axis=0)) / alturas_pesos.std(axis=0)

array([[ 1.36734694, -0.50057491],
       [ 0.34693878,  0.03033787],
       [-1.42176871,  0.94047408],
       [ 0.34693878,  1.39554218],
       [-0.7414966 , -0.87979833],
       [-1.48979592, -1.41071111],
       [-0.60544218,  0.40956129],
       [ 0.68707483,  1.31969749],
       [ 0.00680272,  0.18202724],
       [ 1.50340136, -1.4865558 ]])

O diagrama abaixo mostra como o parâmetro `axis` afeta o calculo da média e do desvio padrão. Mudar esse parâmetro permite você calcular a média por linha, por coluna, ou geral do seus dados.

![Eixos de Agregação](../images/agregacao_numpy.png)

### Pandas

<img src="../images/pandas.jpeg" width=400 align="center">

Pandas é uma biblioteca para Python, desenvolvida para manipular dados tabulares. Essa biblioteca pode não só substituir o excel por uma alternativa mais flexível e programático, mas também é ideal para integras dados com bibliotecas de machine learning, como Sklearn e Tensorflow, que já estão preparados para lidar com dados carregados via pandas.

No Pandas, matrizes são `Dataframes` e as colunas (listas com nomes para as linhas) são chamadas de `Series`.

Para essa parte da aula, usaremos os dados do jogo FIFA 19. Esses dados contém informações sobre todos os jogadores presentes no jogo, de diversos clubes ao redor do mundo. Esses dados podem ser [encontrados no Kaggle](https://www.kaggle.com/karangadiya/fifa19). Apesar de lúdico, esses dados têm muitas das características de dados que encontramos normalmente como data scientists: são dados brutos de indivíduos, de onde queremos extrair informações e padrões.

![fifa](../images/fifa.jpeg)

Nesta aula, usaremos a biblioteca Pandas para extrair informações básicas e responder perguntas sobre nossos dados.

Exemplos de uso deste dataset no futuro: *clustering* dos jogadores, previsão de sua habilidade geral (*Overall*) com base em suas estatísticas, visualização dos dados, etc. Muitas destas aplicações podem ser encontradas em [notebooks do Kaggle](https://www.kaggle.com/karangadiya/fifa19/kernels).

#### Baixando e Carregando dados

In [51]:
data = pd.read_csv("../data/data.csv", index_col=0)

Métodos e atributos bastante comuns:

In [52]:
data.head() #método que mostra as N primeiras linhas; por padrão, N = 5.

ID               Name  Age  \
0  158023           L. Messi   31   
1   20801  Cristiano Ronaldo   33   
2  190871          Neymar Jr   26   
3  193080             De Gea   27   
4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png   

       ...       Composure Marking  StandingTackle SlidingTackle  GKDiving  \
0      ...            96.0    33.0            28.0          26.0       6.0   
1      ...            95.0    28.0            31.0          23.0       7.0   
2      ...            94.0    27.0            24.0          33.0       9.0   
3      ...            68.0    15.0            21.0          13.0      90.0   
4      ...            88.0    68.0            58.0          51.0      15.0   

   GKHandling  GKKicking GKPositioning GKReflexes Release Clause  
0        11.0       15.0          14.0        8.0        €226.5M  
1        11.0       15.0          14.0       11.0        €127.1M  
2         9.0       15.0          15.0       11.0        €228.1M  
3        85.0       87.0          88.0       94.0        €138.6M  
4        13.0        5.0          10.0       13.0        €196.4M  

[5 rows x 88 columns]

In [53]:
data.tail() #método análogo ao .head(), que mostra as N últimas linhas.

ID                Name  Age  \
18202  238813        J. Lundstram   19   
18203  243165  N. Christoffersson   19   
18204  241638           B. Worman   16   
18205  246268      D. Walker-Rice   17   
18206  246269           G. Nugent   16   

                                                Photo Nationality  \
18202  https://cdn.sofifa.org/players/4/19/238813.png     England   
18203  https://cdn.sofifa.org/players/4/19/243165.png      Sweden   
18204  https://cdn.sofifa.org/players/4/19/241638.png     England   
18205  https://cdn.sofifa.org/players/4/19/246268.png     England   
18206  https://cdn.sofifa.org/players/4/19/246269.png     England   

                                      Flag  Overall  Potential  \
18202  https://cdn.sofifa.org/flags/14.png       47         65   
18203  https://cdn.sofifa.org/flags/46.png       47         63   
18204  https://cdn.sofifa.org/flags/14.png       47         67   
18205  https://cdn.sofifa.org/flags/14.png       47         66   
18206  https://cdn.sofifa.org/flags/14.png       46         66   

                   Club                                       Club Logo  \
18202   Crewe Alexandra    https://cdn.sofifa.org/teams/2/light/121.png   
18203    Trelleborgs FF    https://cdn.sofifa.org/teams/2/light/703.png   
18204  Cambridge United   https://cdn.sofifa.org/teams/2/light/1944.png   
18205   Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   
18206   Tranmere Rovers  https://cdn.sofifa.org/teams/2/light/15048.png   

           ...       Composure Marking  StandingTackle SlidingTackle  \
18202      ...            45.0    40.0            48.0          47.0   
18203      ...            42.0    22.0            15.0          19.0   
18204      ...            41.0    32.0            13.0          11.0   
18205      ...            46.0    20.0            25.0          27.0   
18206      ...            43.0    40.0            43.0          50.0   

       GKDiving  GKHandling  GKKicking GKPositioning GKReflexes Release Clause  
18202      10.0        13.0        7.0           8.0        9.0          €143K  
18203      10.0         9.0        9.0           5.0       12.0          €113K  
18204       6.0         5.0       10.0           6.0       13.0          €165K  
18205      14.0         6.0       14.0           8.0        9.0          €143K  
18206      10.0        15.0        9.0          12.0        9.0          €165K  

[5 rows x 88 columns]

In [54]:
data.sample() #método análogo aos dois anterios, mas mostra N linhas aleatórias.

ID           Name  Age  \
13338  235600  B. Al Mutairi   28   

                                                Photo   Nationality  \
13338  https://cdn.sofifa.org/players/4/19/235600.png  Saudi Arabia   

                                       Flag  Overall  Potential      Club  \
13338  https://cdn.sofifa.org/flags/183.png       62         62  Al Batin   

                                             Club Logo      ...        \
13338  https://cdn.sofifa.org/teams/2/light/112990.png      ...         

      Composure Marking  StandingTackle SlidingTackle  GKDiving  GKHandling  \
13338      42.0    58.0            61.0          56.0       7.0         6.0   

       GKKicking GKPositioning GKReflexes Release Clause  
13338       13.0           8.0        9.0          €396K  

[1 rows x 88 columns]

In [55]:
data.shape #atributo que informa a quantidade de linhas e colunas do dataframe

(18207, 88)

In [56]:
data.columns #atributo que informa os nomes das colunas do dataframe

Index(['ID', 'Name', 'Age', 'Photo', 'Nationality', 'Flag', 'Overall',
       'Potential', 'Club', 'Club Logo', 'Value', 'Wage', 'Special',
       'Preferred Foot', 'International Reputation', 'Weak Foot',
       'Skill Moves', 'Work Rate', 'Body Type', 'Real Face', 'Position',
       'Jersey Number', 'Joined', 'Loaned From', 'Contract Valid Until',
       'Height', 'Weight', 'LS', 'ST', 'RS', 'LW', 'LF', 'CF', 'RF', 'RW',
       'LAM', 'CAM', 'RAM', 'LM', 'LCM', 'CM', 'RCM', 'RM', 'LWB', 'LDM',
       'CDM', 'RDM', 'RWB', 'LB', 'LCB', 'CB', 'RCB', 'RB', 'Crossing',
       'Finishing', 'HeadingAccuracy', 'ShortPassing', 'Volleys', 'Dribbling',
       'Curve', 'FKAccuracy', 'LongPassing', 'BallControl', 'Acceleration',
       'SprintSpeed', 'Agility', 'Reactions', 'Balance', 'ShotPower',
       'Jumping', 'Stamina', 'Strength', 'LongShots', 'Aggression',
       'Interceptions', 'Positioning', 'Vision', 'Penalties', 'Composure',
       'Marking', 'StandingTackle', 'SlidingTackle', 'GKDiv

In [57]:
data.describe() #metodo que já nos traz medidas estatísticas da distribuição das colunas

ID           Age       Overall     Potential       Special  \
count   18207.000000  18207.000000  18207.000000  18207.000000  18207.000000   
mean   214298.338606     25.122206     66.238699     71.307299   1597.809908   
std     29965.244204      4.669943      6.908930      6.136496    272.586016   
min        16.000000     16.000000     46.000000     48.000000    731.000000   
25%    200315.500000     21.000000     62.000000     67.000000   1457.000000   
50%    221759.000000     25.000000     66.000000     71.000000   1635.000000   
75%    236529.500000     28.000000     71.000000     75.000000   1787.000000   
max    246620.000000     45.000000     94.000000     95.000000   2346.000000   

       International Reputation     Weak Foot   Skill Moves  Jersey Number  \
count              18159.000000  18159.000000  18159.000000   18147.000000   
mean                   1.113222      2.947299      2.361308      19.546096   
std                    0.394031      0.660456      0.756164      15.947765   
min                    1.000000      1.000000      1.000000       1.000000   
25%                    1.000000      3.000000      2.000000       8.000000   
50%                    1.000000      3.000000      2.000000      17.000000   
75%                    1.000000      3.000000      3.000000      26.000000   
max                    5.000000      5.000000      5.000000      99.000000   

           Crossing      ...          Penalties     Composure       Marking  \
count  18159.000000      ...       18159.000000  18159.000000  18159.000000   
mean      49.734181      ...          48.548598     58.648274     47.281623   
std       18.364524      ...          15.704053     11.436133     19.904397   
min        5.000000      ...           5.000000      3.000000      3.000000   
25%       38.000000      ...          39.000000     51.000000     30.000000   
50%       54.000000      ...          49.000000     60.000000     53.000000   
75%       64.000000      ...          60.000000     67.000000     64.000000   
max       93.000000      ...          92.000000     96.000000     94.000000   

       StandingTackle  SlidingTackle      GKDiving    GKHandling  \
count    18159.000000   18159.000000  18159.000000  18159.000000   
mean        47.697836      45.661435     16.616223     16.391596   
std         21.664004      21.289135     17.695349     16.906900   
min          2.000000       3.000000      1.000000      1.000000   
25%         27.000000      24.000000      8.000000      8.000000   
50%         55.000000      52.000000     11.000000     11.000000   
75%         66.000000      64.000000     14.000000     14.000000   
max         93.000000      91.000000     90.000000     92.000000   

          GKKicking  GKPositioning    GKReflexes  
count  18159.000000   18159.000000  18159.000000  
mean      16.232061      16.388898     16.710887  
std       16.502864      17.034669     17.955119  
min        1.000000       1.000000      1.000000  
25%        8.000000       8.000000      8.000000  
50%       11.000000      11.000000     11.000000  
75%       14.000000      14.000000     14.000000  
max       91.000000      90.000000     94.000000  

[8 rows x 43 columns]

In [58]:
data.info() #método que informa a quantidade de dados não nulos e o tipo do dado para cada coluna

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18207 entries, 0 to 18206
Data columns (total 88 columns):
ID                          18207 non-null int64
Name                        18207 non-null object
Age                         18207 non-null int64
Photo                       18207 non-null object
Nationality                 18207 non-null object
Flag                        18207 non-null object
Overall                     18207 non-null int64
Potential                   18207 non-null int64
Club                        17966 non-null object
Club Logo                   18207 non-null object
Value                       18207 non-null object
Wage                        18207 non-null object
Special                     18207 non-null int64
Preferred Foot              18159 non-null object
International Reputation    18159 non-null float64
Weak Foot                   18159 non-null float64
Skill Moves                 18159 non-null float64
Work Rate                   18159 non-null 

#### Acessando dados de uma tabela

Você consegue acessar uma coluna inteira da tabela usando `df.nomeDaColuna`

In [59]:
data.Name

0                    L. Messi
1           Cristiano Ronaldo
2                   Neymar Jr
3                      De Gea
4                K. De Bruyne
5                   E. Hazard
6                   L. Modrić
7                   L. Suárez
8                Sergio Ramos
9                    J. Oblak
10             R. Lewandowski
11                   T. Kroos
12                   D. Godín
13                David Silva
14                   N. Kanté
15                  P. Dybala
16                    H. Kane
17               A. Griezmann
18              M. ter Stegen
19                T. Courtois
20            Sergio Busquets
21                  E. Cavani
22                   M. Neuer
23                  S. Agüero
24               G. Chiellini
25                  K. Mbappé
26                   M. Salah
27                   Casemiro
28               J. Rodríguez
29                 L. Insigne
                 ...         
18177               R. Roache
18178            L. Wahlstedt
18179     

<div class="alert alert-block alert-warning">
<b>Atenção:</b>

Dessa maneira não é possível acessar colunas cujos nomes contém espaços. Nesses casos, prefira o método `.loc`.

Omitindo o .loc, também conseguimos acessar as colunas utilizando colchetes e aspas (data["Nome"]).
</div>


Além de usar o método acima de acessar, existem outras duas maneiras de acessar dados em pandas: uma usando o método `loc` (mais comum) e outra usando o método `iloc`. Em ambos os casos é possível acessar

* Um elemento da tabela (em geral um `int`, `float` ou `str`)
* Uma coluna (que em pandas é do tipo `pd.Series`)
* Uma sub-tabela (que também é do tipo `pd.DataFrame`, assim como a tabela original)

##### .iloc

O método mais simples é o `iloc`, que deixa você acessar os elementos da tabela usando os números das posições das colunas e das linhas. Isso é chamado de **acesso posicional**. Você sempre usa esse método como `df.iloc[linhas, colunas]`, onde `linhas` e `colunas` podem ser:

* Inteiros, ex. 5.
* Uma lista de inteiros, ex. `[4, 3, 0]`.
* Um objeto slice com inteiros, ex. `1:7`.
* Um array de booleanos, ex `[false, true, false]`.

Por exemplo, se você quiser acessar a linha 0 e a coluna 1, você faz:

In [60]:
data.iloc[0, 1]

'L. Messi'

Agora, usando listas de inteiros, é possível acessar a linha 0 e 1, e as colunas 1 e 2:

In [61]:
data.iloc[[0, 1], [1, 2]]

Name  Age
0           L. Messi   31
1  Cristiano Ronaldo   33

É possível selecionar multiplas linhas e colunas com `slices` usando ranges como `n:m` ao invés de números

In [62]:
data.iloc[0:5, 0:3]

ID               Name  Age
0  158023           L. Messi   31
1   20801  Cristiano Ronaldo   33
2  190871          Neymar Jr   26
3  193080             De Gea   27
4  192985       K. De Bruyne   27

Podemos misturar também!

In [63]:
data.iloc[0:10, 1]

0             L. Messi
1    Cristiano Ronaldo
2            Neymar Jr
3               De Gea
4         K. De Bruyne
5            E. Hazard
6            L. Modrić
7            L. Suárez
8         Sergio Ramos
9             J. Oblak
Name: Name, dtype: object

##### .loc

Também é possível acessar dados partir dos `nomes de linhas e de colunas`, usando o método `loc`. Esse método usa os índices das linhas (que são basicamente os "nomes" das linhas, e os nomes das colunas como forma de acessar os dados da tabela. Esse método é usado como `df.loc[linhas, colunas]`, onde `linhas` e `colunas` podem ser:

* Um label único, como por exemplo `5` ou `'a'`, (note que 5 é interpretado como nome do índice nesse caso, e não como sua posição).
* Uma lista de labels, ex `['a', 'b', 'c']`.
* Um objeto slice com labels como `'a':'f'`.
* Um array de booleanos, ex `[false, true, false]`.

In [64]:
data.loc[1, "Name"]

'Cristiano Ronaldo'

Também é possível usar listas de nomes que você quer acessar:

In [65]:
data.loc[[0,1,3], ["Name", "Age"]]

Name  Age
0           L. Messi   31
1  Cristiano Ronaldo   33
3             De Gea   27

Além de aceitar usar slicing

In [66]:
data.loc[0:5, "ID":"Age"]

ID               Name  Age
0  158023           L. Messi   31
1   20801  Cristiano Ronaldo   33
2  190871          Neymar Jr   26
3  193080             De Gea   27
4  192985       K. De Bruyne   27
5  183277          E. Hazard   27

E também conseguir misturar:

In [67]:
data.loc[0:10, ["Name", "Age"]]

Name  Age
0            L. Messi   31
1   Cristiano Ronaldo   33
2           Neymar Jr   26
3              De Gea   27
4        K. De Bruyne   27
5           E. Hazard   27
6           L. Modrić   32
7           L. Suárez   31
8        Sergio Ramos   32
9            J. Oblak   25
10     R. Lewandowski   29

#### Acesso com array de booleanos

Tanto no caso de `iloc` como de `loc` é possível passar um array de booleanos. Funciona assim, cada booleano, `True` ou `False`, indica se você quer ou não, manter aquela linha/coluna na sua tabela. Por exemplo, se queremos apenas a segunda e a terceira linha da tabela podemos acessar usando um array de booleanos fazendo:

`df.loc[[False, True, True, False, False, False, False, ..., False], :]`

Pode não parecer muito útil a princípio (afinal, poderíamos fazer simplesmente `df.loc[[1,2], :]`, não é mesmo?) mas esta é a principal forma de filtrar dados de uma tabela, como veremos mais pra frente.

<div class="alert alert-block alert-info">
<b>Exercício:</b>

Extraia da tabela o nome e a posição dos dez últimos jogadores da tabela.
</div>

<div class="alert alert-block alert-success">
<b>Solução:</b>
</div>

In [68]:
# usando 'iloc'
data.iloc[-10:, [1, 20]]

Name Position
18197          D. Holland       CM
18198          J. Livesey       GK
18199        M. Baldisimo       CM
18200            J. Young       ST
18201            D. Walsh       RB
18202        J. Lundstram       CM
18203  N. Christoffersson       ST
18204           B. Worman       ST
18205      D. Walker-Rice       RW
18206           G. Nugent       CM

In [69]:
# usando 'loc'
data.loc[18196:, ["Name", "Position"]]
# ou, para não precisar colocar o número exato da linha:
# data.loc[data.shape[0]-10:, ["Name", "Position"]]

Name Position
18196         K. Fujikawa       CM
18197          D. Holland       CM
18198          J. Livesey       GK
18199        M. Baldisimo       CM
18200            J. Young       ST
18201            D. Walsh       RB
18202        J. Lundstram       CM
18203  N. Christoffersson       ST
18204           B. Worman       ST
18205      D. Walker-Rice       RW
18206           G. Nugent       CM

#### Operações lógicas e filtrando dados de uma tabela

Você pode fazer operações lógicas com elementos de uma dataframe. Suponha que vocês queiram ver quais jogadores tem mais de 27 anos:

In [70]:
data.Age >= 28

0         True
1         True
2        False
3        False
4        False
5        False
6         True
7         True
8         True
9        False
10        True
11        True
12        True
13        True
14       False
15       False
16       False
17       False
18       False
19       False
20        True
21        True
22        True
23        True
24        True
25       False
26       False
27       False
28       False
29       False
         ...  
18177    False
18178    False
18179    False
18180    False
18181    False
18182    False
18183     True
18184    False
18185    False
18186    False
18187    False
18188    False
18189    False
18190    False
18191    False
18192    False
18193    False
18194    False
18195    False
18196    False
18197    False
18198    False
18199    False
18200    False
18201    False
18202    False
18203    False
18204    False
18205    False
18206    False
Name: Age, Length: 18207, dtype: bool

Você pode usar operações como `>`, `>=`, `==` para fazer esse tipo de comparação:

In [71]:
data.Club == "FC Barcelona"

0         True
1        False
2        False
3        False
4        False
5        False
6        False
7         True
8        False
9        False
10       False
11       False
12       False
13       False
14       False
15       False
16       False
17       False
18        True
19       False
20        True
21       False
22       False
23       False
24       False
25       False
26       False
27       False
28       False
29       False
         ...  
18177    False
18178    False
18179    False
18180    False
18181    False
18182    False
18183    False
18184    False
18185    False
18186    False
18187    False
18188    False
18189    False
18190    False
18191    False
18192    False
18193    False
18194    False
18195    False
18196    False
18197    False
18198    False
18199    False
18200    False
18201    False
18202    False
18203    False
18204    False
18205    False
18206    False
Name: Club, Length: 18207, dtype: bool

Se você quiser saber quais jogadores são do Barcelona **E** têm mais de 28 anos, pode usar o operador `&` (que é basicamente um `and`, assim como `|` é um `or`):

In [72]:
(data.Age >= 28) & (data.Club == "FC Barcelona")

0         True
1        False
2        False
3        False
4        False
5        False
6        False
7         True
8        False
9        False
10       False
11       False
12       False
13       False
14       False
15       False
16       False
17       False
18       False
19       False
20        True
21       False
22       False
23       False
24       False
25       False
26       False
27       False
28       False
29       False
         ...  
18177    False
18178    False
18179    False
18180    False
18181    False
18182    False
18183    False
18184    False
18185    False
18186    False
18187    False
18188    False
18189    False
18190    False
18191    False
18192    False
18193    False
18194    False
18195    False
18196    False
18197    False
18198    False
18199    False
18200    False
18201    False
18202    False
18203    False
18204    False
18205    False
18206    False
Length: 18207, dtype: bool

Podemos agora usar esses arrays  booleanos para **filtrar nossos dados**, usando `.loc` ou `.iloc`. Se quisermos filtrar a tabela e só manter as linhas dos jogadores do Barcelona, fazemos

In [73]:
data.loc[data.Club == "FC Barcelona", :]  # note que usamos array de booleanos para as linhas e slice para as colunas

ID             Name  Age  \
0      158023         L. Messi   31   
7      176580        L. Suárez   31   
18     192448    M. ter Stegen   26   
20     189511  Sergio Busquets   29   
32     189242         Coutinho   26   
42     205600        S. Umtiti   24   
49     189332       Jordi Alba   29   
53     168651       I. Rakitić   30   
54     152729            Piqué   31   
96     181872         A. Vidal   31   
155    231443       O. Dembélé   21   
192    199564    Sergi Roberto   26   
228    230658           Arthur   21   
236    222737           Malcom   21   
239    220440       C. Lenglet   23   
263    201400          Rafinha   25   
266    199987     J. Cillessen   29   
324    227928    Nélson Semedo   24   
613    205192     Denis Suárez   24   
1007   220253            Munir   22   
1037   157304     T. Vermaelen   32   
3118   208088     Sergi Samper   23   
3684   233113            Aleñá   20   
6000   237156         M. Wagué   19   
6102   242816       Riqui Puig   18   
8284   239249   Oriol Busquets   19   
8289   239250        Abel Ruiz   18   
8857   240915          Miranda   18   
10526  241810            Chumi   19   
10777  237522     Jorge Cuenca   18   
11300  246599    Guillem Jaime   19   
12502  225395          Ezkieta   21   
14286  246139       Iñaki Peña   19   

                                                Photo  Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png    Argentina   
7      https://cdn.sofifa.org/players/4/19/176580.png      Uruguay   
18     https://cdn.sofifa.org/players/4/19/192448.png      Germany   
20     https://cdn.sofifa.org/players/4/19/189511.png        Spain   
32     https://cdn.sofifa.org/players/4/19/189242.png       Brazil   
42     https://cdn.sofifa.org/players/4/19/205600.png       France   
49     https://cdn.sofifa.org/players/4/19/189332.png        Spain   
53     https://cdn.sofifa.org/players/4/19/168651.png      Croatia   
54     https://cdn.sofifa.org/players/4/19/152729.png        Spain   
96     https://cdn.sofifa.org/players/4/19/181872.png        Chile   
155    https://cdn.sofifa.org/players/4/19/231443.png       France   
192    https://cdn.sofifa.org/players/4/19/199564.png        Spain   
228    https://cdn.sofifa.org/players/4/19/230658.png       Brazil   
236    https://cdn.sofifa.org/players/4/19/222737.png       Brazil   
239    https://cdn.sofifa.org/players/4/19/220440.png       France   
263    https://cdn.sofifa.org/players/4/19/201400.png       Brazil   
266    https://cdn.sofifa.org/players/4/19/199987.png  Netherlands   
324    https://cdn.sofifa.org/players/4/19/227928.png     Portugal   
613    https://cdn.sofifa.org/players/4/19/205192.png        Spain   
1007   https://cdn.sofifa.org/players/4/19/220253.png        Spain   
1037   https://cdn.sofifa.org/players/4/19/157304.png      Belgium   
3118   https://cdn.sofifa.org/players/4/19/208088.png        Spain   
3684   https://cdn.sofifa.org/players/4/19/233113.png        Spain   
6000   https://cdn.sofifa.org/players/4/19/237156.png      Senegal   
6102   https://cdn.sofifa.org/players/4/19/242816.png        Spain   
8284   https://cdn.sofifa.org/players/4/19/239249.png        Spain   
8289   https://cdn.sofifa.org/players/4/19/239250.png        Spain   
8857   https://cdn.sofifa.org/players/4/19/240915.png        Spain   
10526  https://cdn.sofifa.org/players/4/19/241810.png        Spain   
10777  https://cdn.sofifa.org/players/4/19/237522.png        Spain   
11300  https://cdn.sofifa.org/players/4/19/246599.png        Spain   
12502  https://cdn.sofifa.org/players/4/19/225395.png        Spain   
14286  https://cdn.sofifa.org/players/4/19/246139.png        Spain   

                                       Flag  Overall  Potential          Club  \
0       https://cdn.sofifa.org/flags/52.png       94         94  FC Barcelona   
7       https://cdn.sofifa.org/flags/60.png       91         91  FC Barcelona   
18      https://cdn.sofifa.org/flags/21.png       89        

<div class="alert alert-block alert-info">
<b>Exercício:</b>

Filtre a tabela, mantendo somente os jogadores do barcelona **E** de nacionalidade espanhola.
</div>

<div class="alert alert-block alert-success">
<b>Solução:</b>
</div>

In [74]:
data.loc[(data.Club == "FC Barcelona") & (data.Nationality == "Spain"), :]

ID             Name  Age  \
20     189511  Sergio Busquets   29   
49     189332       Jordi Alba   29   
54     152729            Piqué   31   
192    199564    Sergi Roberto   26   
613    205192     Denis Suárez   24   
1007   220253            Munir   22   
3118   208088     Sergi Samper   23   
3684   233113            Aleñá   20   
6102   242816       Riqui Puig   18   
8284   239249   Oriol Busquets   19   
8289   239250        Abel Ruiz   18   
8857   240915          Miranda   18   
10526  241810            Chumi   19   
10777  237522     Jorge Cuenca   18   
11300  246599    Guillem Jaime   19   
12502  225395          Ezkieta   21   
14286  246139       Iñaki Peña   19   

                                                Photo Nationality  \
20     https://cdn.sofifa.org/players/4/19/189511.png       Spain   
49     https://cdn.sofifa.org/players/4/19/189332.png       Spain   
54     https://cdn.sofifa.org/players/4/19/152729.png       Spain   
192    https://cdn.sofifa.org/players/4/19/199564.png       Spain   
613    https://cdn.sofifa.org/players/4/19/205192.png       Spain   
1007   https://cdn.sofifa.org/players/4/19/220253.png       Spain   
3118   https://cdn.sofifa.org/players/4/19/208088.png       Spain   
3684   https://cdn.sofifa.org/players/4/19/233113.png       Spain   
6102   https://cdn.sofifa.org/players/4/19/242816.png       Spain   
8284   https://cdn.sofifa.org/players/4/19/239249.png       Spain   
8289   https://cdn.sofifa.org/players/4/19/239250.png       Spain   
8857   https://cdn.sofifa.org/players/4/19/240915.png       Spain   
10526  https://cdn.sofifa.org/players/4/19/241810.png       Spain   
10777  https://cdn.sofifa.org/players/4/19/237522.png       Spain   
11300  https://cdn.sofifa.org/players/4/19/246599.png       Spain   
12502  https://cdn.sofifa.org/players/4/19/225395.png       Spain   
14286  https://cdn.sofifa.org/players/4/19/246139.png       Spain   

                                      Flag  Overall  Potential          Club  \
20     https://cdn.sofifa.org/flags/45.png       89         89  FC Barcelona   
49     https://cdn.sofifa.org/flags/45.png       87         87  FC Barcelona   
54     https://cdn.sofifa.org/flags/45.png       87         87  FC Barcelona   
192    https://cdn.sofifa.org/flags/45.png       83         86  FC Barcelona   
613    https://cdn.sofifa.org/flags/45.png       79         86  FC Barcelona   
1007   https://cdn.sofifa.org/flags/45.png       77         83  FC Barcelona   
3118   https://cdn.sofifa.org/flags/45.png       73         80  FC Barcelona   
3684   https://cdn.sofifa.org/flags/45.png       72         85  FC Barcelona   
6102   https://cdn.sofifa.org/flags/45.png       69         89  FC Barcelona   
8284   https://cdn.sofifa.org/flags/45.png       67         82  FC Barcelona   
8289   https://cdn.sofifa.org/flags/45.png       67         83  FC Barcelona   
8857   https://cdn.sofifa.org/flags/45.png       66         81  FC Barcelona   
10526  https://cdn.sofifa.org/flags/45.png       65         82  FC Barcelona   
10777  https://cdn.sofifa.org/flags/45.png       65         78  FC Barcelona   
11300  https://cdn.sofifa.org/flags/45.png       64         80  FC Barcelona   
12502  https://cdn.sofifa.org/flags/45.png       63         75  FC Barcelona   
14286  https://cdn.sofifa.org/flags/45.png       61         78  FC Barcelona   

                                          Club Logo      ...       Composure  \
20     https://cdn.sofifa.org/teams/2/light/241.png      ...            90.0   
49     https://cdn.sofifa.org/teams/2/light/241.png      ...            79.0   
54     https://cdn.sofifa.org/teams/2/light/241.png      ...            86.0   
192    https://cdn.sofifa.org/teams/2/light/241.png      ...            76.0   
613    https://cdn.sofifa.org/teams/2/light/241.png      ...            70.0   
1007   https://cdn.sofifa.org/teams/2/light/241.png      ...            76.0   
3118   https://cdn.sofifa.org/teams/2/light/241.png      ...    

#### Outras operações com pandas

Você também pode usar métodos de soma, média, desvio padrão para colunas (ou seja, objetos do tipo `pd.Series`) do pandas. Isso é muito simples, já que você precisa apenas usar métodos como `series.sum()`. Por exemplo, qual é a média de idade dos jogadores do Fifa? Podemos responder isso fazendo:

In [75]:
data.Age.mean()

25.122205745043114

Outros métodos são `.sum()`, `.min()`, `.max()`, `.stddev()`, mas diversos outros estão disponíveis na [documentação do pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html).

<div class="alert alert-block alert-info">
<b>Exercício:</b>

Encontre a menor idade dentre os jogadores do Barcelona. Qual (ou quais) jogadores do barcelona possuem essa idade?
</div>

<div class="alert alert-block alert-success">
<b>Solução:</b>
</div>

In [76]:
age_min_barc = data.loc[data.Club == "FC Barcelona", "Age"].min()
age_min_barc

18

In [77]:
data.loc[(data.Club == "FC Barcelona") & (data.Age == age_min_barc), :]

ID          Name  Age  \
6102   242816    Riqui Puig   18   
8289   239250     Abel Ruiz   18   
8857   240915       Miranda   18   
10777  237522  Jorge Cuenca   18   

                                                Photo Nationality  \
6102   https://cdn.sofifa.org/players/4/19/242816.png       Spain   
8289   https://cdn.sofifa.org/players/4/19/239250.png       Spain   
8857   https://cdn.sofifa.org/players/4/19/240915.png       Spain   
10777  https://cdn.sofifa.org/players/4/19/237522.png       Spain   

                                      Flag  Overall  Potential          Club  \
6102   https://cdn.sofifa.org/flags/45.png       69         89  FC Barcelona   
8289   https://cdn.sofifa.org/flags/45.png       67         83  FC Barcelona   
8857   https://cdn.sofifa.org/flags/45.png       66         81  FC Barcelona   
10777  https://cdn.sofifa.org/flags/45.png       65         78  FC Barcelona   

                                          Club Logo      ...       Composure  \
6102   https://cdn.sofifa.org/teams/2/light/241.png      ...            70.0   
8289   https://cdn.sofifa.org/teams/2/light/241.png      ...            57.0   
8857   https://cdn.sofifa.org/teams/2/light/241.png      ...            59.0   
10777  https://cdn.sofifa.org/teams/2/light/241.png      ...            57.0   

      Marking  StandingTackle SlidingTackle  GKDiving  GKHandling  GKKicking  \
6102     53.0            57.0          53.0      12.0        15.0       14.0   
8289     45.0            12.0          12.0       6.0        10.0       10.0   
8857     64.0            69.0          67.0      11.0         8.0       12.0   
10777    62.0            66.0          63.0      10.0        12.0        5.0   

      GKPositioning GKReflexes Release Clause  
6102           12.0        9.0          €6.5M  
8289           10.0        8.0          €4.2M  
8857            9.0        6.0          €3.2M  
10777          14.0       12.0          €2.3M  

[4 rows x 88 columns]

<div class="alert alert-block alert-info">
<b>Exercício:</b>

A habilidade do jogador, de maneira geral, está contida na coluna `Overall`. Qual time possui a maior média de habilidade, Barecelona ou Real Madrid? Qual dos dois times tem o melhor jogador?
</div>

<div class="alert alert-block alert-success">
<b>Solução:</b>
</div>

In [78]:
data.loc[data.Club == "FC Barcelona", "Overall"].mean()

78.03030303030303

In [79]:
data.loc[data.Club == "Real Madrid", "Overall"].mean()

78.24242424242425

O 'Overall' do Real Madrid é um pouco maior, mas qual deles tem o melhor jogador?

In [80]:
data.loc[data.Club == "FC Barcelona", "Overall"].max()

94

In [81]:
data.loc[data.Club == "Real Madrid", "Overall"].max()

91

O Barcelona!

Podemos fazer como no exercício anterior para verificar quais jogadores tem o valor de Overall igual a esses encontrados:

In [82]:
melhor_barca = data.loc[data.Club == "FC Barcelona", "Overall"].max()

In [83]:
data.loc[(data.Club == "FC Barcelona") & (data.Overall == melhor_barca)]

ID      Name  Age                                           Photo  \
0  158023  L. Messi   31  https://cdn.sofifa.org/players/4/19/158023.png   

  Nationality                                 Flag  Overall  Potential  \
0   Argentina  https://cdn.sofifa.org/flags/52.png       94         94   

           Club                                     Club Logo      ...        \
0  FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png      ...         

  Composure Marking  StandingTackle SlidingTackle  GKDiving  GKHandling  \
0      96.0    33.0            28.0          26.0       6.0        11.0   

   GKKicking GKPositioning GKReflexes Release Clause  
0       15.0          14.0        8.0        €226.5M  

[1 rows x 88 columns]

#### Ordenando os dados

Nem sempre a tabela está ordenada da maneira ideal para resolver o seu problema. Por exemplo, você pode querer ver os jogadores do FIFA ordenados por idade. Para isso, usaremos o método `.sort_values()`:

In [84]:
data.sort_values(by="Age")

ID                Name  Age  \
18206  246269           G. Nugent   16   
17743  244752           J. Olstad   16   
13293  246594         H. Massengo   16   
16081  241552         J. Italiano   16   
18166  243621            N. Ayéva   16   
17712  245470            K. Broda   16   
17354  242240         L. D'Arrigo   16   
13567  246419      Y. Verschaeren   16   
16927  243646           B. Nygren   16   
18018  243722         B. O'Gorman   16   
17091  243269            A. Doğan   16   
17115  245341          C. Bassett   16   
15793  241650            J. Lahne   16   
15746  243169         Y. Begraoui   16   
17175  243353            B. Mumba   16   
18003  239594        J. Imbrechts   16   
18106  243685           J. Cleary   16   
17177  242074            R. Gómez   16   
17200  245171        H. Andersson   16   
17976  243891           M. Larsen   16   
16418  242531         J. Kitolano   16   
17263  245485     P. Samiec-Talar   16   
18124  245808        G. Hollywood   16   
15363  245015           Y. Roemer   16   
12828  246465            Guerrero   16   
12496  245616      Pelayo Morilla   16   
18162  243866        T. Gundelund   16   
18044  246109             K. Lara   16   
17921  246601          J. Rowland   16   
16254  244728           S. Steijn   16   
...       ...                 ...  ...   
1952   161468          P. Guiñazú   39   
5040   142670            R. Braña   39   
6861    49212          J. Speroni   39   
4062   159520        N. Fernández   39   
5101   105308           J. Gillet   39   
7427   146636           José Juan   39   
14920   53352           D. Bulman   39   
11456  152690  D. Konstantopoulos   39   
10356   49511            F. Kippe   40   
12453  214372             W. Díaz   40   
864    153260              Hilton   40   
16264   28157     P. van der Vlag   40   
14129  156480         Y. Nakazawa   40   
17752  100899         S. Phillips   40   
41       1179           G. Buffon   40   
16596  219344        A. Al Basisi   40   
2821   232543          S. Bertoli   40   
3550    29552         S. Nakamura   40   
4187   156483        C. Lucchetti   40   
9484   176540         B. Castillo   40   
1294    14907         A. Bizzarri   40   
1120   156092           J. Villar   41   
7225   142998            C. Muñoz   41   
15426   18745            M. Tyler   41   
4228     3665            B. Nivet   41   
12192  208927        H. Sulaimani   41   
10545  140183         S. Narazaki   42   
18183   53748       K. Pilkington   44   
17726   51963           T. Warner   44   
4741   140029            O. Pérez   45   

                                                Photo          Nationality  \
18206  https://cdn.sofifa.org/players/4/19/246269.png              England   
17743  https://cdn.sofifa.org/players/4/19/244752.png               Norway   
13293  https://cdn.sofifa.org/players/4/19/246594.png               France   
16081  https://cdn.sofifa.org/players/4/19/241552.png            Australia   
18166  https://cdn.sofifa.org/players/4/19/243621.png               Sweden   
17712  https://cdn.sofifa.org/players/4/19/245470.png               Poland   
17354  https://cdn.sofifa.org/players/4/19/242240.png            Australia   
13567  https://cdn.sofifa.org/players/4/19/246419.png              Belgium   
16927  https://cdn.sofifa.org/players/4/19/243646.png               Sweden   
18018  https://cdn.sofifa.org/players/4/19/243722.png  Republic of Ireland   
17091  https://cdn.sofifa.org/players/4/19/243269.png               Turkey   
17115  https://cdn.sofifa.org/players/4/19/245341.png        United States   
15793  https://cdn.sofifa.org/players/4/19/241650.png               Sweden   
15746  https://cdn.sofifa.org/players/4/19/243169.png               France   
17175  https://cdn.sofifa.org/players/4/19/243353.png              England   
18003  https://cdn.sofifa.org/players/4/19/239594.png               Sweden   
18106  https://cdn.sofifa.org/players/4/19/243685.png  Republic of Ireland   
17

Se quisermos inverter a ordem, basta usar o parâmetro `ascending`:

In [85]:
data.sort_values(by="Age", ascending=False)

ID             Name  Age  \
4741   140029         O. Pérez   45   
18183   53748    K. Pilkington   44   
17726   51963        T. Warner   44   
10545  140183      S. Narazaki   42   
7225   142998         C. Muñoz   41   
1120   156092        J. Villar   41   
12192  208927     H. Sulaimani   41   
15426   18745         M. Tyler   41   
4228     3665         B. Nivet   41   
10356   49511         F. Kippe   40   
16264   28157  P. van der Vlag   40   
9484   176540      B. Castillo   40   
4187   156483     C. Lucchetti   40   
2821   232543       S. Bertoli   40   
3550    29552      S. Nakamura   40   
1294    14907      A. Bizzarri   40   
14129  156480      Y. Nakazawa   40   
41       1179        G. Buffon   40   
864    153260           Hilton   40   
12453  214372          W. Díaz   40   
17752  100899      S. Phillips   40   
16596  219344     A. Al Basisi   40   
2225    25420       C. Pizarro   39   
1952   161468       P. Guiñazú   39   
4640    16097       N. Rimando   39   
6861    49212       J. Speroni   39   
7427   146636        José Juan   39   
5608   156662        A. Bastía   39   
4062   159520     N. Fernández   39   
10534  140180     M. Ogasawara   39   
...       ...              ...  ...   
16418  242531      J. Kitolano   16   
16544  240562       D. Adshead   16   
16927  243646        B. Nygren   16   
17091  243269         A. Doğan   16   
17115  245341       C. Bassett   16   
17175  243353         B. Mumba   16   
17177  242074         R. Gómez   16   
17200  245171     H. Andersson   16   
17263  245485  P. Samiec-Talar   16   
15793  241650         J. Lahne   16   
15746  243169      Y. Begraoui   16   
17808  245123     A. Mahlonoko   16   
18018  243722      B. O'Gorman   16   
13567  246419   Y. Verschaeren   16   
18124  245808     G. Hollywood   16   
13293  246594      H. Massengo   16   
18106  243685        J. Cleary   16   
12828  246465         Guerrero   16   
18044  246109          K. Lara   16   
12496  245616   Pelayo Morilla   16   
18003  239594     J. Imbrechts   16   
17841  244925   F. Tauchhammer   16   
17976  243891        M. Larsen   16   
11732  244403         A. Taoui   16   
11457  241266      W. Geubbels   16   
17921  246601       J. Rowland   16   
17887  246009         M. Tilio   16   
15363  245015        Y. Roemer   16   
17881  242165         R. Hauge   16   
18206  246269        G. Nugent   16   

                                                Photo          Nationality  \
4741   https://cdn.sofifa.org/players/4/19/140029.png               Mexico   
18183   https://cdn.sofifa.org/players/4/19/53748.png              England   
17726   https://cdn.sofifa.org/players/4/19/51963.png    Trinidad & Tobago   
10545  https://cdn.sofifa.org/players/4/19/140183.png                Japan   
7225   https://cdn.sofifa.org/players/4/19/142998.png            Argentina   
1120   https://cdn.sofifa.org/players/4/19/156092.png             Paraguay   
12192  https://cdn.sofifa.org/players/4/19/208927.png         Saudi Arabia   
15426   https://cdn.sofifa.org/players/4/19/18745.png              England   
4228     https://cdn.sofifa.org/players/4/19/3665.png               France   
10356   https://cdn.sofifa.org/players/4/19/49511.png               Norway   
16264   https://cdn.sofifa.org/players/4/19/28157.png          Netherlands   
9484   https://cdn.sofifa.org/players/4/19/176540.png             Colombia   
4187   https://cdn.sofifa.org/players/4/19/156483.png            Argentina   
2821   https://cdn.sofifa.org/players/4/19/232543.png            Argentina   
3550    https://cdn.sofifa.org/players/4/19/29552.png                Japan   
1294    https://cdn.sofifa.org/players/4/19/14907.png            Argentina   
14129  https://cdn.sofifa.org/players/4/19/156480.png                Japan   
41       https://cdn.sofifa.org/players/4/19/1179.png                Italy   
864    https://cdn.sofifa.org/players/4/19/153260.png               Brazil   
12453  https://cdn.sofifa.org/pl

Podemos ordenar por duas colunas. Quando você faz isso, o pandas ordena pela primeira, e quando há empate, ordena pela segunda:

In [86]:
data.sort_values(by=["Age", "Overall"], ascending=False)

ID             Name  Age  \
4741   140029         O. Pérez   45   
17726   51963        T. Warner   44   
18183   53748    K. Pilkington   44   
10545  140183      S. Narazaki   42   
1120   156092        J. Villar   41   
4228     3665         B. Nivet   41   
7225   142998         C. Muñoz   41   
12192  208927     H. Sulaimani   41   
15426   18745         M. Tyler   41   
41       1179        G. Buffon   40   
864    153260           Hilton   40   
1294    14907      A. Bizzarri   40   
2821   232543       S. Bertoli   40   
3550    29552      S. Nakamura   40   
4187   156483     C. Lucchetti   40   
9484   176540      B. Castillo   40   
10356   49511         F. Kippe   40   
12453  214372          W. Díaz   40   
14129  156480      Y. Nakazawa   40   
16264   28157  P. van der Vlag   40   
16596  219344     A. Al Basisi   40   
17752  100899      S. Phillips   40   
554     49031    S. Sorrentino   39   
1952   161468       P. Guiñazú   39   
2225    25420       C. Pizarro   39   
2706   110638        Cifuentes   39   
3128     7647      D. Dainelli   39   
3420   149571    S. Pellissier   39   
3603    16254        T. Howard   39   
4062   159520     N. Fernández   39   
...       ...              ...  ...   
16544  240562       D. Adshead   16   
16927  243646        B. Nygren   16   
17091  243269         A. Doğan   16   
17115  245341       C. Bassett   16   
17175  243353         B. Mumba   16   
17177  242074         R. Gómez   16   
17200  245171     H. Andersson   16   
17263  245485  P. Samiec-Talar   16   
17354  242240      L. D'Arrigo   16   
17712  245470         K. Broda   16   
17743  244752        J. Olstad   16   
17751  245537         E. Ceide   16   
17753  245539         B. Waine   16   
17757  246315         L. Smyth   16   
17776  244553   M. Köstenbauer   16   
17808  245123     A. Mahlonoko   16   
17841  244925   F. Tauchhammer   16   
17881  242165         R. Hauge   16   
17887  246009         M. Tilio   16   
17921  246601       J. Rowland   16   
17976  243891        M. Larsen   16   
18003  239594     J. Imbrechts   16   
18018  243722      B. O'Gorman   16   
18044  246109          K. Lara   16   
18106  243685        J. Cleary   16   
18124  245808     G. Hollywood   16   
18162  243866     T. Gundelund   16   
18166  243621         N. Ayéva   16   
18204  241638        B. Worman   16   
18206  246269        G. Nugent   16   

                                                Photo          Nationality  \
4741   https://cdn.sofifa.org/players/4/19/140029.png               Mexico   
17726   https://cdn.sofifa.org/players/4/19/51963.png    Trinidad & Tobago   
18183   https://cdn.sofifa.org/players/4/19/53748.png              England   
10545  https://cdn.sofifa.org/players/4/19/140183.png                Japan   
1120   https://cdn.sofifa.org/players/4/19/156092.png             Paraguay   
4228     https://cdn.sofifa.org/players/4/19/3665.png               France   
7225   https://cdn.sofifa.org/players/4/19/142998.png            Argentina   
12192  https://cdn.sofifa.org/players/4/19/208927.png         Saudi Arabia   
15426   https://cdn.sofifa.org/players/4/19/18745.png              England   
41       https://cdn.sofifa.org/players/4/19/1179.png                Italy   
864    https://cdn.sofifa.org/players/4/19/153260.png               Brazil   
1294    https://cdn.sofifa.org/players/4/19/14907.png            Argentina   
2821   https://cdn.sofifa.org/players/4/19/232543.png            Argentina   
3550    https://cdn.sofifa.org/players/4/19/29552.png                Japan   
4187   https://cdn.sofifa.org/players/4/19/156483.png            Argentina   
9484   https://cdn.sofifa.org/players/4/19/176540.png             Colombia   
10356   https://cdn.sofifa.org/players/4/19/49511.png               Norway   
12453  https://cdn.sofifa.org/players/4/19/214372.png             Colombia   
14129  https://cdn.sofifa.org/players/4/19/156480.png                Japan   
16264   https://cdn.sofifa.org/p

##### Alterando dados e criando colunas novas

Em pandas, podemos manipular os dados de diversas maneiras. A operações mais comuns (como `+`, `-`, `==` etc), podem ser usadas diretamente, como qualquer outro objeto Python. No caso de uma coluna de um DataFrame, a operação é feita **termo a termo**:

In [87]:
data["Age"] + 10

0        41
1        43
2        36
3        37
4        37
5        37
6        42
7        41
8        42
9        35
10       39
11       38
12       42
13       42
14       37
15       34
16       34
17       37
18       36
19       36
20       39
21       41
22       42
23       40
24       43
25       29
26       36
27       36
28       36
29       37
         ..
18177    28
18178    28
18179    27
18180    32
18181    27
18182    28
18183    54
18184    28
18185    29
18186    30
18187    29
18188    27
18189    28
18190    28
18191    28
18192    28
18193    28
18194    28
18195    28
18196    29
18197    28
18198    28
18199    28
18200    28
18201    28
18202    29
18203    29
18204    26
18205    27
18206    26
Name: Age, Length: 18207, dtype: int64

Podemos colocar os nossos resultados em colunas novas do DataFrame se desejarmos. Para fazer isso, simplesmente use um *assignment* do Python, assim:

In [88]:
data["Age in 10 years"] = data["Age"] + 10

No entanto, muitas vezes precisamos fazer uso de funções mais complicadas. Se conseguirmos fazer essa operação como uma função `f(x)` em Python, podemos aplicar essa função em cada elemento da coluna usando o método `.apply(f)`. Por exemplo:

In [89]:
def funcao_complicada(overall):
    return (overall / 10) * ((overall + 2) ** 2)

data.Overall.apply(funcao_complicada)

0        86630.4
1        86630.4
2        81291.2
3        78705.9
4        78705.9
5        78705.9
6        78705.9
7        78705.9
8        78705.9
9        76176.0
10       76176.0
11       76176.0
12       76176.0
13       76176.0
14       73700.9
15       73700.9
16       73700.9
17       73700.9
18       73700.9
19       73700.9
20       73700.9
21       73700.9
22       73700.9
23       73700.9
24       73700.9
25       71280.0
26       71280.0
27       71280.0
28       71280.0
29       71280.0
          ...   
18177    12000.0
18178    12000.0
18179    12000.0
18180    12000.0
18181    12000.0
18182    12000.0
18183    12000.0
18184    12000.0
18185    12000.0
18186    11284.7
18187    11284.7
18188    11284.7
18189    11284.7
18190    11284.7
18191    11284.7
18192    11284.7
18193    11284.7
18194    11284.7
18195    11284.7
18196    11284.7
18197    11284.7
18198    11284.7
18199    11284.7
18200    11284.7
18201    11284.7
18202    11284.7
18203    11284.7
18204    11284

##### Agrupando dados e tirando estatísticas

Muitas vezes precisamos agrupar e tirar informações de cada grupo na nossa tabela. Por exemplo, podemos querer saber a idade média por clube, o *Overall* médio por país, o número de clubes por país, o número de jogadores por país etc. Para extrair todas essas informações, precisamos agrupar os dados e tirar alguma informação de cada grupo. A biblioteca pandas facilita esse trabalho e dá maneiras simples de responder todas essas perguntas.

A agregação mais simples que podemos fazer em Pandas é a *quantidade de vezes que cada elemento aparece em uma coluna*, usando o método `.value_counts()`. Por exemplo, para encontrar a quantidade de jogadores por país, fazemos:

In [90]:
data.Nationality.value_counts()

England                 1662
Germany                 1198
Spain                   1072
Argentina                937
France                   914
Brazil                   827
Italy                    702
Colombia                 618
Japan                    478
Netherlands              453
Sweden                   397
China PR                 392
Chile                    391
Republic of Ireland      368
Mexico                   366
United States            353
Poland                   350
Norway                   341
Saudi Arabia             340
Denmark                  336
Korea Republic           335
Portugal                 322
Turkey                   303
Austria                  298
Scotland                 286
Belgium                  260
Australia                236
Switzerland              220
Uruguay                  149
Senegal                  130
                        ... 
Hong Kong                  2
Uzbekistan                 2
Nicaragua                  2
Bermuda       

Para agregar estatísticas mais complicadas, precisamos usar o método `.groupby()`. Por exemplo, para encontrar a média de *Overall* por clube, fazemos assim: fazemos um grupo por cada *Club* usando `.groupby("Club")`, depois pegamos a coluna `Overall` e tiramos a a média usando `.mean()`.

In [91]:
data.groupby("Club").Overall.mean()

Club
 SSV Jahn Regensburg        65.586207
1. FC Heidenheim 1846       65.750000
1. FC Kaiserslautern        63.384615
1. FC Köln                  70.785714
1. FC Magdeburg             65.615385
1. FC Nürnberg              68.827586
1. FC Union Berlin          68.321429
1. FSV Mainz 05             70.843750
AC Ajaccio                  65.043478
AC Horsens                  60.640000
AD Alcorcón                 67.413793
ADO Den Haag                66.678571
AEK Athens                  70.214286
AFC Wimbledon               60.461538
AIK                         65.074074
AJ Auxerre                  66.296296
AS Béziers                  62.038462
AS Monaco                   72.939394
AS Nancy Lorraine           64.666667
AS Saint-Étienne            70.875000
AZ Alkmaar                  70.000000
Aalborg BK                  62.037037
Aarhus GF                   61.407407
Aberdeen                    64.333333
Accrington Stanley          61.178571
Adelaide United             61.400000
Ajax   

Podemos pegar tirar a média de mais uma coluna, usando `[col1, col2]` para acessá-las. Por exemplo, para tirar a média de `Age` e de `Overall` para cada `Club`, fazemos:

In [92]:
data.groupby("Club")["Overall", "Age"].mean()

Overall        Age
Club                                          
 SSV Jahn Regensburg      65.586207  25.655172
1. FC Heidenheim 1846     65.750000  24.000000
1. FC Kaiserslautern      63.384615  23.846154
1. FC Köln                70.785714  24.321429
1. FC Magdeburg           65.615385  24.692308
1. FC Nürnberg            68.827586  23.793103
1. FC Union Berlin        68.321429  25.250000
1. FSV Mainz 05           70.843750  23.687500
AC Ajaccio                65.043478  27.043478
AC Horsens                60.640000  25.000000
AD Alcorcón               67.413793  26.896552
ADO Den Haag              66.678571  24.678571
AEK Athens                70.214286  24.535714
AFC Wimbledon             60.461538  23.615385
AIK                       65.074074  26.074074
AJ Auxerre                66.296296  25.185185
AS Béziers                62.038462  25.769231
AS Monaco                 72.939394  23.060606
AS Nancy Lorraine         64.666667  24.933333
AS Saint-Étienne          70.875000  25.416667
AZ Alkmaar                70.000000  23.066667
Aalborg BK                62.037037  23.370370
Aarhus GF                 61.407407  23.777778
Aberdeen                  64.333333  24.148148
Accrington Stanley        61.178571  24.821429
Adelaide United           61.400000  24.360000
Ajax                      73.633333  22.300000
Akhisar Belediyespor      69.115385  29.269231
Al Ahli                   65.700000  25.933333
Al Batin                  59.833333  25.100000
...                             ...        ...
Vitesse                   68.464286  24.464286
Vitória                   70.400000  30.000000
Vitória Guimarães         72.333333  24.566667
Vitória de Setúbal        68.892857  25.392857
Vålerenga Fotball         61.769231  23.192308
Vélez Sarsfield           65.250000  23.214286
Waasland-Beveren          64.714286  22.500000
Walsall                   63.370370  23.814815
Waterford FC              57.200000  23.880000
Watford                   72.344828  25.482759
Wellington Phoenix        62.190476  25.380952
West Bromwich Albion      69.533333  25.800000
West Ham United           73.093750  25.437500
Western Sydney Wanderers  61.840000  24.080000
Wigan Athletic            66.566667  24.533333
Willem II                 65.285714  22.678571
Wisła Kraków              62.461538  24.500000
Wisła Płock               61.038462  24.115385
Wolfsberger AC            61.538462  24.615385
Wolverhampton Wanderers   68.818182  22.848485
Wycombe Wanderers         63.720000  27.400000
Yeni Malatyaspor          65.566667  25.866667
Yeovil Town               60.357143  24.250000
Yokohama F. Marinos       61.533333  25.033333
Zagłębie Lubin            62.629630  25.703704
Zagłębie Sosnowiec        60.760000  26.240000
Çaykur Rizespor           66.900000  25.433333
Örebro SK                 60.481481  24.037037
Östersunds FK             63.545455  23.863636
Śląsk Wrocław             62.200000  25.960000

[651 rows x 2 columns]

<div class="alert alert-block alert-warning">
<b>Atenção:</b>

Note que a coluna que você agrupou, nesse caso `Club`, vira o índice das linhas no resultado final.
</div>

<div class="alert alert-block alert-info">
<b>Exercício:</b>

Potencial não realizado é a diferença entre as colunas `Potential` (potencial que o jogador tem de chegar em termos de habilidade) e `Overall` (habilidade atual do jogador). Encontre o time de **maior média de potencial não realizado**. 
</div>

<div class="alert alert-block alert-success">
<b>Solução:</b>
</div>

In [93]:
data["Unrealized Potential"] = data.Potential - data.Overall

In [94]:
data.groupby("Club")["Unrealized Potential"].mean().sort_values(ascending=False)

Club
FC Nordsjælland              11.000000
Stabæk Fotball               10.888889
Envigado FC                  10.178571
Derry City                   10.111111
FC Sochaux-Montbéliard       10.000000
Crewe Alexandra               9.464286
LOSC Lille                    9.320000
Bohemian FC                   9.240000
Barnsley                      9.142857
Bray Wanderers                8.956522
FC Groningen                  8.807692
Wolverhampton Wanderers       8.757576
Sligo Rovers                  8.736842
Hamburger SV                  8.733333
Shamrock Rovers               8.590909
Colchester United             8.392857
Club Atlético Banfield        8.333333
Vejle Boldklub                8.333333
Atlético Huila                8.321429
Olympique Lyonnais            8.275862
FC Red Bull Salzburg          8.222222
Vélez Sarsfield               8.214286
CD Huachipato                 8.074074
Club Atlético Talleres        8.071429
AS Monaco                     8.000000
FC Admira Wacker Möd

Podemos criar um novo DataFrame a partir dessa informação agregada que geramos agrupamento por Club:

In [95]:
df_clubs = pd.DataFrame(data.groupby("Club")["Unrealized Potential"].mean())
df_clubs.head()

Unrealized Potential
Club                                       
 SSV Jahn Regensburg               3.724138
1. FC Heidenheim 1846              6.178571
1. FC Kaiserslautern               6.500000
1. FC Köln                         5.785714
1. FC Magdeburg                    4.730769

Se quisermos tirar o Club do index (nome de linha), podemos usar a função `reset_index`:

In [96]:
df_clubs = df_clubs.reset_index()
df_clubs.head()

Club  Unrealized Potential
0    SSV Jahn Regensburg              3.724138
1  1. FC Heidenheim 1846              6.178571
2   1. FC Kaiserslautern              6.500000
3             1. FC Köln              5.785714
4        1. FC Magdeburg              4.730769

Agora podemos ordenar pela coluna de potencial não realizado:

In [97]:
df_clubs_ordered = df_clubs.sort_values(by='Unrealized Potential', ascending = False)
df_clubs_ordered.head()

Club  Unrealized Potential
231         FC Nordsjælland             11.000000
559          Stabæk Fotball             10.888889
202             Envigado FC             10.178571
188              Derry City             10.111111
237  FC Sochaux-Montbéliard             10.000000

E podemos também selecionar o clube que tem o maior potencial não realizado médio:

In [98]:
df_clubs_ordered.iloc[0]

Club                    FC Nordsjælland
Unrealized Potential                 11
Name: 231, dtype: object

In [99]:
df_clubs_ordered[:1] # Também podemos fazer a seleção dessa forma, semelhante a acessos a listas:

Club  Unrealized Potential
231  FC Nordsjælland                  11.0

#### Dados ausentes ou nulos

Em muitos casos, os dados que um data scientist vai trabalhar não estão em perfeitas condições. Em muitos casos, várias colunas possuem, por exemplo, dados ausentes, seja por falhas no processo de capptura, por esses dados estarem indisponíveis, ou por esse dado não fazer sentido para aquele caso (por exemplo, na linha de um jogador novato, a coluna de "Clube anterior" pode estar vazia pois aquele é seu primeiro clube).

Para investigar se cada elemento de uma coluna é nulo ou não, podemos usar os métodos `.isnull()` ou `.notnull()`. Por exemplo, para vez quais elementos da coluna `Club` estão nulos:

In [100]:
data.Club.isnull()

0        False
1        False
2        False
3        False
4        False
5        False
6        False
7        False
8        False
9        False
10       False
11       False
12       False
13       False
14       False
15       False
16       False
17       False
18       False
19       False
20       False
21       False
22       False
23       False
24       False
25       False
26       False
27       False
28       False
29       False
         ...  
18177    False
18178    False
18179    False
18180    False
18181    False
18182    False
18183    False
18184    False
18185    False
18186    False
18187    False
18188    False
18189    False
18190    False
18191    False
18192    False
18193    False
18194    False
18195    False
18196    False
18197    False
18198    False
18199    False
18200    False
18201    False
18202    False
18203    False
18204    False
18205    False
18206    False
Name: Club, Length: 18207, dtype: bool

Mas quantos deles são nulos? Para isso, basta somar todos os `True` que temos no resultado acima: 

In [101]:
data.Club.isnull().sum()

241

Então 241 jogadores estão sem clube! Podemos agora querer preencher esses valores nulos com a string `"No Club"`. Para fazer isso, usamos o método `.fillna()`:

In [102]:
data["Filled Club"] = data.Club.fillna("No Club")

In [103]:
data["Filled Club"].isnull().sum()

0

Note que a coluna `Filled Club` não tem mais valores nulos.

Outra operação muito comum em trabalho de data science é jogar fora colunas que tem algum valor nulo. Podemos fazer isso usando o método `dropna()`, assim:

In [104]:
data.dropna(axis="columns")

ID                  Name  Age  \
0      158023              L. Messi   31   
1       20801     Cristiano Ronaldo   33   
2      190871             Neymar Jr   26   
3      193080                De Gea   27   
4      192985          K. De Bruyne   27   
5      183277             E. Hazard   27   
6      177003             L. Modrić   32   
7      176580             L. Suárez   31   
8      155862          Sergio Ramos   32   
9      200389              J. Oblak   25   
10     188545        R. Lewandowski   29   
11     182521              T. Kroos   28   
12     182493              D. Godín   32   
13     168542           David Silva   32   
14     215914              N. Kanté   27   
15     211110             P. Dybala   24   
16     202126               H. Kane   24   
17     194765          A. Griezmann   27   
18     192448         M. ter Stegen   26   
19     192119           T. Courtois   26   
20     189511       Sergio Busquets   29   
21     179813             E. Cavani   31   
22     167495              M. Neuer   32   
23     153079             S. Agüero   30   
24     138956          G. Chiellini   33   
25     231747             K. Mbappé   19   
26     209331              M. Salah   26   
27     200145              Casemiro   26   
28     198710          J. Rodríguez   26   
29     198219            L. Insigne   27   
...       ...                   ...  ...   
18177  238550             R. Roache   18   
18178  243158          L. Wahlstedt   18   
18179  246243           J. Williams   17   
18180  221669              M. Hurst   22   
18181  245734              C. Maher   17   
18182  246001               Y. Góez   18   
18183   53748         K. Pilkington   44   
18184  241657             D. Horton   18   
18185  243961              E. Tweed   19   
18186  240917          Zhang Yufeng   20   
18187  240158             C. Ehlich   19   
18188  240927            L. Collins   17   
18189  240160            A. Kaltner   18   
18190  245569            L. Watkins   18   
18191  245570  J. Norville-Williams   18   
18192  245571             S. Squire   18   
18193  244823            N. Fuentes   18   
18194  245862              J. Milli   18   
18195  243582            S. Griffin   18   
18196  238477           K. Fujikawa   19   
18197  246167            D. Holland   18   
18198  242844            J. Livesey   18   
18199  244677          M. Baldisimo   18   
18200  231381              J. Young   18   
18201  243413              D. Walsh   18   
18202  238813          J. Lundstram   19   
18203  243165    N. Christoffersson   19   
18204  241638             B. Worman   16   
18205  246268        D. Walker-Rice   17   
18206  246269             G. Nugent   16   

                                                Photo          Nationality  \
0      https://cdn.sofifa.org/players/4/19/158023.png            Argentina   
1       https://cdn.sofifa.org/players/4/19/20801.png             Portugal   
2      https://cdn.sofifa.org/players/4/19/190871.png               Brazil   
3      https://cdn.sofifa.org/players/4/19/193080.png                Spain   
4      https://cdn.sofifa.org/players/4/19/192985.png              Belgium   
5      https://cdn.sofifa.org/players/4/19/183277.png              Belgium   
6      https://cdn.sofifa.org/players/4/19/177003.png              Croatia   
7      https://cdn.sofifa.org/players/4/19/176580.png              Uruguay   
8      https://cdn.sofifa.org/players/4/19/155862.png                Spain   
9      https://cdn.sofifa.org/players/4/19/200389.png             Slovenia   
10     https://cdn.sofifa.org/players/4/19/188545.png               Poland   
11     https://cdn.sofifa.org/players/4/19/182521.png              Germany   
12     https://cdn.sofifa.org/players/4/19/182493.png              Uruguay   
13     https://cdn.sofifa.org/players/4/19/168542.png                Spain   
14     https://cdn.sofifa.org/players/4/19/215914.png               France   
15     https://cdn.sofifa.org/play

Veja que só sobraram 14 colunas. Se quisermos jogar todas as linhas que tenham algum valor nulo, basta colocar o parâmetro `axis="rows"` no método `.dropna()`

## Pandas ou numpy?

É importante saber alguns prós e contras do Numpy e Pandas para quando vamos escolher a melhor ferramenta para realizar alguma tarefa. 

Algumas dicas:

* Dados tabulares -> preferência a Pandas
* problema por estritamente um problema de algebra linear (como decomposição de matrizes ou manipulação de vetores) -> escolha Numpy 
* Performance: numpy é mais simples e sua simplicidade resulta em um código mais otimizado (com diversas partes internas escritas em C ao invés de Python). Se a performance das suas operações no pandas é um problema, troque para Numpy para fazer os cálculos mais pesados.
* **Usem SEMPRE a documentação e a comunidade (como Stack Overflow), pois essas são as principais vantagens de utilizar Python e essas bibliotecas mais "famosas".**

### Esse material foi produzido com base no material de Paulo Roberto de Oliveira Castro.